doc/ws.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "WS" "1" "September 2017" "www.complex-networks.net" "www.complex-networks.net"
.
.SH "NAME"
\fBws\fR \- Create a small\-world graph using the Watts\-Strogatz model
.
.SH "SYNOPSIS"
\fBws\fR \fIN\fR \fIm\fR \fIp\fR [SHOW]
.
.SH "DESCRIPTION"
\fBws\fR creates a small\-world undirected graph with \'N\' nodes using the Watts\-Strogatz small\-world network model\. The nodes are initially placed around a circle and each node is connected to its \'m\' closest neighbours on either side\. Then, each edge is rewired (independently) with probability \'p\'\. The program prints on output the edge\-list of the resulting graph\.
.
.SH "PARAMETERS"
.
.TP
\fIN\fR
Number of nodes in the final graph\.
.
.TP
\fIm\fR
Number of neighbours on the circle graph to which a node is connected on either side (i\.e\., each node will initially have \'2m\' edges)\.
.
.TP
\fIp\fR
Edge rewiring probability\.
.
.TP
SHOW
If the fourth (optional) parameter is equal to \fBSHOW\fR, the program will print on STDERR the number of edges rewired while constructing the graph\.
.
.SH "EXAMPLES"
The command:
.
.IP "" 4
.
.nf

      $ ws 100 3 0\.1
      0 95
      0 2
      0 3
      1 2
      1 24
      1 4
      2 3
      \.\.\.\.
      99 0
      99 56
      99 2
      $
.
.fi
.
.IP "" 0
.
.P
creates a Watts\-Strogatz small\-world network with \'N=100\' nodes and \'K=m*N=300\' edges, where each edge is independently rewired with probability \'p=0\.1\'\. The output of \fBws\fR is the edge\-list of the resulting graph\. Notice that in this case the first edge od node \fB0\fR has been rewired and connected to node \fB95\fR\. In order to be useful, such edge\-list should be saved into a file\. The following command:
.
.IP "" 4
.
.nf

      $ ws 100 300 0\.1 SHOW > ws_100_3_0\.1\.net
      Rewired edges: 32
      $
.
.fi
.
.IP "" 0
.
.P
will save the resulting graph in the file \fBws_100_3_0\.1\.net\fR\. Notice the usage of the symbol ">" to redirect the output of the program to a file\. The line printed on output reports the number of rewired edges, since \fBSHOW\fR was specified as a fourth parameter\.
.
.SH "SEE ALSO"
er_A(1), er_B(1)
.
.SH "REFERENCES"
.
.IP "\(bu" 4
D\. J\. Watts and S\. H\. Strogatz\. "Collective dynamics of \'small\-world\' networks"\. Nature 393 (1998), 440–442\.
.
.IP "\(bu" 4
V\. Latora, V\. Nicosia, G\. Russo, "Complex Networks: Principles, Methods and Applications", Chapter 4, Cambridge University Press (2017)
.
.IP "" 0
.
.SH "AUTHORS"
(c) Vincenzo \'KatolaZ\' Nicosia 2009\-2017 \fB<v\.nicosia@qmul\.ac\.uk>\fR\.


debian/changelog
netbunch (1.0-1) unstable; urgency=medium

  * Initial release 

 -- Vincenzo (KatolaZ) Nicosia <katolaz@freaknet.org>  Thu, 28 Sep 2017 06:46:47 +0100


src/ws/ws.c
/**
 *  This program is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see
 *  <http://www.gnu.org/licenses/>.
 *
 *  (c) Vincenzo Nicosia 2009-2017 -- <v.nicosia@qmul.ac.uk>
 * 
 *  This file is part of NetBunch, a package for complex network
 *  analysis and modelling. For more information please visit:
 *
 *             http://www.complex-networks.net/
 *
 *  If you use this software, please add a reference to 
 *
 *               V. Latora, V. Nicosia, G. Russo             
 *   "Complex Networks: Principles, Methods and Applications"
 *              Cambridge University Press (2017) 
 *                ISBN: 9781107103184
 *
 ***********************************************************************
 *
 *  This program creates a network using the Watts-Strogatz
 *  small-world network model.
 *
 *  References: 
 *
 *  D. J. Watts and S. H. Strogatz. "Collective dynamics of
 *  'small-world' networks".  Nature 393 (1998), 440–442.
 *  
 *  
 */


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "utils.h"

void usage(char *argv[]){
  printf("********************************************************************\n"
         "**                                                                **\n"
         "**                        -*-     ws     -*-                      **\n"
         "**                                                                **\n"
         "**    Create a graph with 'N' nodes, using the Watts-Strogatz     **\n"
         "**    small-world network model. The algorithm starts from a      **\n"
         "**    circle graph with 'N' nodes and m*N edges, i.e. by putting  **\n"
         "**    the 'N' nodes around a circle, so that each node is         **\n"
         "**    connected to the 'm' closest nodes in the circle on each    **\n"
         "**    direction. Then, each edge is rewired at random with        **\n"
         "**    probability equal to 'p'.                                   **\n"
         "**                                                                **\n"
         "**    The output is an edge list in the format:                   **\n"
         "**                                                                **\n"
         "**                            I_1 J_1                             **\n"
         "**                            I_2 J_2                             **\n"
         "**                            I_3 J_3                             **\n"
         "**                            ... ...                             **\n"
         "**                            I_K J_K                             **\n"
         "**                                                                **\n"
         "**    If 'SHOW' is specified as a fourth parameter, the program   **\n"
         "**    prints on STDERR the number of edges that were actually     **\n"
         "**    rewired.                                                    **\n"
         "**                                                                **\n"
         "********************************************************************\n"
         " This is Free Software - You can use and distribute it under \n"
         " the terms of the GNU General Public License, version 3 or later\n\n"
         " (c) Vincenzo Nicosia 2010-2017 (v.nicosia@qmul.ac.uk)\n\n"
         "********************************************************************\n\n"
         );
  printf("Usage: %s <N> <m> <p> [SHOW]\n\n" , argv[0]);
}



/**
 *
 * This function checks if j is a neighbour of i, looking into the 
 * m positions of J starting at i*m 
 * 
 * In practice, the assumption is that J contains the m neighbours of
 * node 0 in the first m positions, then the m neighbours of node 1,
 * and to forth
 *
 */

int __ws_is_neigh(unsigned int i, unsigned int j, unsigned int *J, unsigned int m){
  
  int k;

  for(k = i*m; k< (i+1)*m; k++){
    if (J[k] == j)
      return 1;
  }
  return 0;
}


int create_circle(unsigned int N, unsigned int m, unsigned int **J){

  int K;
  int i, j, l;
  
  K = N *m;
  
  *J = malloc(K * sizeof(unsigned int));

  K = 0;
  
  for(i=0; i<N; i++){
    for(j=0; j < m; j++){
      l = (i + j + 1) % N;
      (*J)[K] = l;
      K += 1;
    }
  }
  return K;
}



int ws(unsigned int *J, unsigned int N, unsigned int m, double p){

  unsigned int i, j, l, l1, num_rewire;
  
  double xi;

  num_rewire = 0;
  
  for(i=0; i<N; i++){
    for(j=0; j<m; j++){
      l = (i+j) % N;
      xi = 1.0 * rand() / RAND_MAX;
      if (xi < p){
        l1 = (int)(rand() % N);
        if( (l1 != i) &&
            (l1 != l) &&
            !(__ws_is_neigh(i, l1, J, m))){
          /* replace (i,l) with (i,l1) */
          J[m * i + j] = l1;
          num_rewire += 1;
        }
        else{
          /* do nothing */
        }
      }
    }
  }
  return num_rewire;
}


void dump_edges(unsigned int *J, unsigned int N, unsigned int m){
  
  int i, j;
  
  for(i=0; i<N; i++){
    for(j=0; j<m; j++){
      printf("%d %d\n", i, J[(m*i)+j]);
    }
  }
}


int main(int argc, char *argv[]){

  int N, m, num_rewire;
  unsigned int  *J;
  double p;

  if(argc < 4){
    usage(argv);
    exit(1);
  }

  J = NULL;
  
  N = atoi(argv[1]);
  m = atoi(argv[2]);
  p = atof(argv[3]);

  create_circle(N, m, &J);
  
  
  srand(time(NULL));
  
  num_rewire = ws(J, N, m, p);
  if (argc > 4 && !my_strcasecmp("SHOW", argv[4])){
    fprintf(stderr, "Rewired edges: %d\n", num_rewire);
  }
  dump_edges(J, N, m);
  free(J);
}
